### DISCUSSION TITLE
A potential bug in the future or is this the expected behavior?

### DISCUSSION THREAD

**Discussion by muyangye** [apache/streampipes#1872] [Good to know]
I am not sure if this is a bug or the expected behavior so I am not opening an issue but a discussion. Also, there's currently no processor that satisfies the "How to Reproduce" requirements. However, I still want to report it as users may modify processors on their own and encounter this.

### What happened?
I modified the PropertyScope of **BooleanOperatorProcessor** from PropertyScope.NONE to PropertyScope.MEASUREMENT_PROPERTY:
![e0d8a1f72a0936a0a38a1a77a08fd0c](https://github.com/apache/streampipes/assets/76769570/690652c7-aeb2-4797-af76-9c077bdf28be)
So now b4, whose PropertyScope is PROPERTY_SCOPE.HEADER_PROPERTY, is not shown in the UI:
![80b4a16200255c673f81afbdede7e22](https://github.com/apache/streampipes/assets/76769570/3d4f7e09-3d21-4926-82fb-d7fd9ecfbbf0)
But its value is always **true**:
![d0b42b32f7a8a91437c2dd723376706](https://github.com/apache/streampipes/assets/76769570/a4dbbbca-07ef-4d97-bdbf-df5aa273e3b7)
After I clicked **SELECT ALL** in the UI and connected to a DataLake, the result is:
![dca3d141ff4db8f6b1806c0460e1e05](https://github.com/apache/streampipes/assets/76769570/b2a51e62-eaf6-47f2-a6ca-eca41208877d)
Clearly this is not what I expected since the UI only shows b0, b1, and b2. However, it seems that the result takes into account of b4. As evident not only by the above picture but also my debug message that prints all input streams:
![1693950191187](https://github.com/apache/streampipes/assets/76769570/3c673bf5-d19c-49f0-8989-e6f043abb96f)
However, if I didn't click **SELECT ALL** but instead clicked b0, b1, and b2 one by one, the result becomes correct:
![a614bdecdddeeecca9687304d913165](https://github.com/apache/streampipes/assets/76769570/add51311-a96d-4d1f-90a1-84fe7dc56037)

### How to Reproduce?
Requirements:

1. Any processor that uses Nary Property Mapping (so that there's a **SELECT ALL** button in the UI)
2. PropertyScope set to anything that is NOT PropertyScope.NONE
3. There exists a property in the input stream that satisfies EpRequirements but is not the same PropertyScope as in the code
4. Click **SELECT ALL** in the UI

Again, as far as I can tell, there's no processor that satisfies both 1 and 2. So I am wondering if the developer is expected to set PropertyScope to PropertyScope.NONE when the mapping is Nary. If not, I am pretty sure this will come up in the future when there is such a combination of PropertyMapping, PropertyScope, and input stream described above.

### Are you willing to submit a PR?
If everyone thinks this is indeed a bug instead of the expected behavior, I will submit an issue and try to investigate the root cause.

**Comment by dominikriemer** [comment]
Hi @muyangye, thanks for this great description of the bug you encountered!
I have not yet found some time to reproduce the error, but I'm pretty sure it's a bug in the UI.
There is a slider in the pipeline element configuration dialog (`Show only recommended`), which basically hides fields based on the property scope (e.g., when a pipeline element requires a property scope `measurement`, all fields that have another property scope are hidden. If you deactivate the slider, you should see the whole list of properties.
My guess is that the UI widget for displaying the NaryMappingProperty selects all fields by default, and then hides fields based on the property scope, leading to the behaviour you described.

Do you think this could be the cause? Then we can create an issue and should be able to quickly solve the bug. If not, I'll try to reproduce the error in the next day or two :-)

Thanks again for the great feedback!

**Reply by muyangye** [reply]
Hi @dominikriemer, thanks for the quick answer!
You are totally correct. I didn't notice there is a `Show only recommended` slider before, after disabling it b4 showed up. I have looked into the Typescript code that filters properties based on property scopes so I guess this isn't really a bug since that's the expected logic. That being said, The fact that NaryMappingProperty has `Show only recommended` default to true is somewhat confusing since it initially selects all properties and `SELECT ALL` also selects all properties, including the ones that are hidden by the `Show only recommended` slider. I think the slider should be changed to default false for NaryMappingProperty. I will submit an issue and discuss my intended approach more closely in the issue :)
